### PR TITLE
feat(enterprise): add ready for handshake support to pgcoord

### PR DESCRIPTION
**Pull request by coadler**

### CODE REVIEW

**Comment by coadler** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/coder/coder/12935?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#12935** <a href="https://app.graphite.dev/github/pr/coder/coder/12935?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#12786** <a href="https://app.graphite.dev/github/pr/coder/coder/12786?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @coadler and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: enterprise/tailnet/handshaker_test.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+	require.NoError(t, err)
+	defer coord1.Close()
+	coord2, err := tailnet.NewPGCoord(ctx, logger.Named("coord2"), ps, store)
+	require.NoError(t, err)
+	defer coord2.Close()
+
+	clientID, agentID := uuid.New(), uuid.New()
+
+	cReq, cRes := coord1.Coordinate(ctx, clientID, clientID.String(), agpl.ClientCoordinateeAuth{AgentID: agentID})
+	aReq, aRes := coord2.Coordinate(ctx, agentID, agentID.String(), agpl.AgentCoordinateeAuth{ID: agentID})
```
Comment (spikecurtis) [inline enterprise/tailnet/handshaker_test.go:39]: this test will be fewer lines and more readable if you use `tailnet/test/peer.go` framework.  Just need to extend it to include ReadyForHandshake.

File: enterprise/tailnet/handshaker_test.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+	"cdr.dev/slog"
+	"cdr.dev/slog/sloggers/slogtest"
+	"github.com/coder/coder/v2/coderd/database/dbtestutil"
+	"github.com/coder/coder/v2/enterprise/tailnet"
+	agpl "github.com/coder/coder/v2/tailnet"
+	"github.com/coder/coder/v2/tailnet/proto"
+	"github.com/coder/coder/v2/testutil"
+)
+
+func TestPGCoordinatorDual_ReadyForHandshake_OK(t *testing.T) {
```
Comment (spikecurtis) [inline enterprise/tailnet/handshaker_test.go:20]: nice

File: coderd/database/db.go [hunk 3]
```diff
@@ -103,7 +103,7 @@ func (q *sqlQuerier) InTx(function func(Store) error, txOpts *sql.TxOptions) err
 				// Transaction succeeded.
 				return nil
 			}
-			if err != nil && !IsSerializedError(err) {
+			if !IsSerializedError(err) {
```
Comment (spikecurtis) [inline coderd/database/db.go:106]: why is this change needed?

File: coderd/database/queries/tailnet.sql [hunk 4]
```diff
@@ -207,6 +207,12 @@ FROM tailnet_tunnels
 INNER JOIN tailnet_peers ON tailnet_tunnels.src_id = tailnet_peers.id
 WHERE tailnet_tunnels.dst_id = $1;
 
+-- name: PublishReadyForHandshake :exec
+SELECT pg_notify(
```
[the remark this answers is not in the document]
Comment (coadler) [reply coderd/database/queries/tailnet.sql:211]: It made it easy to mock the call, and in the no-permission test make sure it wasn't being called

[the remark this answers is not in the document]
Comment (coadler) [reply coderd/database/queries/tailnet.sql:211]: not necessary anymore with your changes tho :+1:

File: enterprise/tailnet/handshaker.go [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+}
+
+type handshaker struct {
+	ctx           context.Context
+	logger        slog.Logger
+	coordinatorID uuid.UUID
+	pubsub        pubsub.Pubsub
+	updates       <-chan readyForHandshake
+
+	workQ *workQ[hKey]
```
Comment (spikecurtis) [inline enterprise/tailnet/handshaker.go:31]: I think this is unused

File: enterprise/tailnet/handshaker.go [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+		select {
+		case <-t.ctx.Done():
+			t.logger.Debug(t.ctx, "handshaker worker exiting", slog.Error(t.ctx.Err()))
+			return
+
+		case rfh := <-t.updates:
+			err := t.pubsub.Publish(eventReadyForHandshake, []byte(fmt.Sprintf(
+				"%s,%s", rfh.dst.String(), rfh.src.String(),
+			)))
+			if err != nil && !database.IsQueryCanceledError(err) {
```
Comment (spikecurtis) [inline enterprise/tailnet/handshaker.go:76]: I don't think you can get a query canceled error from `Publish()`

File: enterprise/tailnet/pgcoord.go [hunk 7]
```diff
@@ -1247,7 +1323,7 @@ type querierWorkKey struct {
 }
 
 type queueKey interface {
-	bKey | tKey | querierWorkKey
+	bKey | tKey | querierWorkKey | hKey
```
Comment (spikecurtis) [inline enterprise/tailnet/pgcoord.go:1326]: I think hKey itself can be dropped and the fields directly put into `readyForHandshake` since we don't queue up this kind of work anymore.

File: enterprise/tailnet/handshaker_test.go [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+	}
+	store, ps := dbtestutil.NewDB(t)
+	ctx, cancel := context.WithTimeout(context.Background(), testutil.WaitSuperLong)
+	defer cancel()
+	logger := slogtest.Make(t, nil).Leveled(slog.LevelDebug)
+	coord1, err := tailnet.NewPGCoord(ctx, logger.Named("coord1"), ps, store)
+	require.NoError(t, err)
+	defer coord1.Close()
+
+	agpltest.ReadyForHandshakeTest(ctx, t, coord1)
```
Comment (spikecurtis) [inline enterprise/tailnet/handshaker_test.go:30]: Very nice structure & code reuse

**Review by spikecurtis — APPROVED** [review]
LGTM!

I made some inline comments about a few leftovers that can get cleaned up, but I don't need to review again.
